### DISCUSSION TITLE
GeometryBuilder::build_* methods not consuming the builder

### DISCUSSION THREAD

**Discussion by VictorKoenders** [17cupsofcoffee/tetra#253] [Q&A]
I ran into an oddity just now where I was using 2 geometrybuilders, and I turned them into a mesh with `build_mesh`:

```rust
let mut tile_builder = GeometryBuilder::new();
let mut tile_selector_builder = GeometryBuilder::new();
// render stuff to the 2 builders
let tile = tile_builder.build_mesh(ctx)?;
let tile_selector = tile_builder.build_mesh(ctx)?;
```

I noticed that this is valid code. Based on the method name `build_mesh` I assumed that the builder would be consumed, which would've prevented the small bug I have above.

According to the [Rust naming guidelines](https://doc.rust-lang.org/1.0.0/style/style/naming/conversions.html) the common keyword for consuming a struct are `.as_`, `.to_` and `.into_`. `.build_` is not one of those keywords.

My question is if it makes sense for other people that `.build_*` would consume the builder, or if they think it's fine as-is.

(for a backwards compatible change we could introduce a `.finish_mesh` and a `.finish_buffer` method instead)

**Comment by 17cupsofcoffee** [comment]
The main reason I made them non-consuming is so you can `clear` them and re-use them after you're done uploading to the GPU - `GeometryBuilder` is basically just two `Vec`s in a trenchcoat so it seemed a bit wasteful to just throw those allocations away 😅

I am interested to see if other people think the naming is confusing though! The current names came about because I originally just had one `build` method (which I think is fairly normal for builders whether they're consuming or not?), but then I realized I wanted to be able to build buffers as well, so I split it into `build_mesh` and `build_buffers`. It felt like it made sense at the time, but I also didn't think too hard about it.
